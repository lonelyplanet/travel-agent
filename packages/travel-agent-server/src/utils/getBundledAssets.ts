import * as path from "path";
import * as fs from "fs";

export default function getBundledAssets() {
  const manifest = require(path.join(process.cwd(), "public", "assets", "manifest.json"));

  return manifest;
}
