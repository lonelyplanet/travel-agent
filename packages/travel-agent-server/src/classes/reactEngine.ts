import * as path from "path";
import * as React from "react";
// import * as pretty from "pretty";
import { renderToStaticMarkup, renderToString } from "react-dom/server";
import getBundledAssets from "../utils/getBundledAssets";
import Helmet, { HelmetData } from "react-helmet";

export interface ILayoutOptions {
  [key: string]: any;
  assets: { [key: string]: string },
  body: string,
  initialState: string,
  head: HelmetData;
}

export interface IReactEngineOptions {
  layout?: string;
  isProdEnv?: boolean;
}

export function getInitialState(options): string {
  const locals = { ...options };
  delete locals.webpackStats;
  delete locals.settings;
  delete locals._locals;
  delete locals.cache;

  return JSON.stringify(locals);
}

export function getMarkupWithDoctype(markup: string): string {
  return `<!DOCTYPE html>\
  ${markup}`;
};

export function generateMarkup(req: NodeRequire, filePath: string, options, defaultLayout: string): string {
  const Component = req(`${filePath}`).default;
  const markup = renderToString(React.createElement(Component, options, null));

  if (options.layout === false || filePath.indexOf("error") > -1) {
    return markup;
  }

  const Layout = req(path.join(process.cwd(), options.layout || defaultLayout)).default;
  const assets = getBundledAssets();
  const initialState = getInitialState(options);
  const head = Helmet.renderStatic();

  // pretty(markup) for prod?
  return renderToStaticMarkup(
    React.createElement(Layout, {
      ...options,
      assets,
      body: markup,
      initialState,
      head,
    }),
  );
};

export default (
  engineOptions: IReactEngineOptions = { layout: "app/layout", isProdEnv: false },
  req = require
) => {
  function ReactEngine(filePath: string, options, callback) {
    try {
      const markup = getMarkupWithDoctype(generateMarkup(req, filePath, options, engineOptions.layout));
      return callback(null, markup);

    } catch (e) {
      return callback(e);

    } finally {
      // For prod, remove all files from the module cache that are in the view folder.
      if (engineOptions.isProdEnv) {
        const moduleDetectRegEx = /.*\.tsx$/;
        Object.keys(req.cache).forEach((module) => {
          if (moduleDetectRegEx.test(req.cache[module].filename)) {
            delete req.cache[module];
          }
        });
      }
    }
  }

  return ReactEngine;
};
