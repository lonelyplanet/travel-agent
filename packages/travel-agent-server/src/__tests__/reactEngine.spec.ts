const getBundledAssetsDefault = jest.fn();
jest.mock("../utils/getBundledAssets", () => ({ default: getBundledAssetsDefault }));
import getBundledAssets from "../utils/getBundledAssets";

const isProdEnvDefault = jest.fn();
jest.mock("../utils/isProdEnv", () => ({ default: isProdEnvDefault }));
import isProdEnv from "../utils/isProdEnv";

import reactEngine, { getInitialState, generateMarkup, getMarkupWithDoctype } from "../classes/reactEngine";

const TestComponent = require("./fixtures/component");
const TestLayout = require("./fixtures/layout");

describe("reactEngine", () => {
  const filePath = "app/modules/home/components";
  const options = {
    message: "This is my message",
    webpackStats: "stats",
    settings: {},
    _locals: "",
    cache: {},
  };
  const requireCache = {
    "ts-file.ts": { filename: "ts-file.ts" },
    "view-file.tsx": { filename: "view-file.tsx" }
  };
  let callback;
  let req;

  beforeEach(() => {
    getBundledAssetsDefault.mockReturnValue({ "vendor.js": "/assets/vendor.js" });
    isProdEnvDefault.mockReturnValue(false);
    callback = jest.fn();
    req = jest.fn<NodeRequire>(() => ({ cache: "stuff" }))
      .mockImplementationOnce((id: string) => TestComponent)
      .mockImplementationOnce((id: string) => TestLayout);
    req.cache = requireCache;
  });

  afterEach(() => {
    getBundledAssetsDefault.mockReset();
    callback.mockReset();
    req.mockReset();
  });

  describe("getInitialState", () => {
    it("should handle an empty object", () => {
      const result = getInitialState({});
      expect(result).toEqual("{}");
    });

    it("should return the JSON string of the correct object", () => {
      const result = getInitialState(options);
      expect(result).toEqual("{\"message\":\"This is my message\"}");
    });
  });

  describe("getMarkupWithDoctype", () => {
    it("should add the doctype to the html string", () => {
      const markup = getMarkupWithDoctype("<html><body>Hello</body></html>");
      expect(markup).toMatchSnapshot();
    });
  });

  describe("generateMarkup", () => {
    it("should generate the correct markup", () => {
      const markup = generateMarkup(req, filePath, options, "app/layout");
      expect(markup).toMatchSnapshot();
    });

    it("should generate the correct markup if the layout is false", () => {
      const extendedOptions = { ...options, layout: false };
      const markup = generateMarkup(req, filePath, extendedOptions, "app/layout");
      expect(markup).toMatchSnapshot();
    });
  });

  describe("reactEngine default", () => {
    it("should call the callback with the correct markup", () => {
      reactEngine({ layout: "app/layout" }, req)(filePath, options, callback);

      expect(callback).toHaveBeenCalled();
      expect(callback.mock.calls[0]).toMatchSnapshot();
      expect(req.cache).toEqual(requireCache);
    });

    it("should hande errors when trying to generate markup", () => {
      const error = new Error("oh snap");
      getBundledAssetsDefault.mockImplementation(() => {throw error});
      reactEngine({ layout: "app/layout" }, req)(filePath, options, callback);

      expect(callback).toHaveBeenCalledWith(error);
    });

    it("should remove cached views from require in a prod environment", () => {
      isProdEnvDefault.mockReturnValue(true);
      reactEngine({ layout: "app/layout" }, req)(filePath, options, callback);

      expect(callback).toHaveBeenCalled();
      expect(req.cache).toEqual({ "ts-file.ts": { filename: "ts-file.ts" } });
    });
  });
});
