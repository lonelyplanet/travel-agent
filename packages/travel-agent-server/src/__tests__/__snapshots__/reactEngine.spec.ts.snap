// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reactEngine generateMarkup should generate the correct markup 1`] = `"<html><body><div id=\\"app\\"><div class=\\"my-component\\" data-reactroot=\\"\\">This is my message</div></div><script type=\\"application/json\\" id=\\"initialState\\">{\\"message\\":\\"This is my message\\"}</script><script src=\\"/assets/vendor.js\\"></script></body></html>"`;

exports[`reactEngine generateMarkup should generate the correct markup if the layout is false 1`] = `"<div class=\\"my-component\\" data-reactroot=\\"\\">This is my message</div>"`;

exports[`reactEngine getMarkupWithDoctype should add the doctype to the html string 1`] = `"<!DOCTYPE html>  <html><body>Hello</body></html>"`;

exports[`reactEngine reactEngine default should call the callback with the correct markup 1`] = `
Array [
  null,
  "<!DOCTYPE html>  <html><body><div id=\\"app\\"><div class=\\"my-component\\" data-reactroot=\\"\\">This is my message</div></div><script type=\\"application/json\\" id=\\"initialState\\">{\\"message\\":\\"This is my message\\"}</script><script src=\\"/assets/vendor.js\\"></script></body></html>",
]
`;
